nemo/collections/nlp/models/language_modeling/megatron_lm_adapter_encoder_decoder_model.py

```diff
@@ -0,0 +1,166 @@
+                self.adapter_cfg = ParallelLinearAdapterConfig(
+                    in_features=cfg.encoder.hidden_size,
+                    dim=cfg.adapter_tuning.adapter_dim,
+                    norm_position=cfg.adapter_tuning.get('norm_position', 'pre'),
+                    norm_type=cfg.adapter_tuning.get('norm_type', 'mixedfusedlayernorm'),
+                    column_init_method=cfg.adapter_tuning.get('column_init_method', 'xavier'),
+                    row_init_method=cfg.adapter_tuning.get('row_init_method', 'zero'),
+                    dropout=cfg.adapter_tuning.adapter_dropout,
+                )
+            elif cfg.adapter_tuning.type == 'linear_adapter':
+                self.adapter_cfg = LinearAdapterConfig(
+                    in_features=cfg.hidden_size,
+                    dim=cfg.adapter_tuning.adapter_dim,
+                    norm_position=cfg.adapter_tuning.get('norm_position', 'pre'),
+                    dropout=cfg.adapter_tuning.adapter_dropout,
+                )
+
+            # override checkpoint saving
+            self.state_dict = self._state_dict
+            self.load_state_dict = self._load_adapters_weights
+            self.add_enable_adapters()
+
+    @classmethod
+    def list_available_models(cls):
+        pass
+
+    def add_enable_adapters(self):
+        logging.info(f'Before adding adapters:\n{self.summarize()}')
+        print(self.adapter_cfg)
+        for _, module in self.enc_dec_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    module.add_adapter(name=adapter_key, cfg=self.adapter_cfg)
+        logging.info(f'After adding adapters:\n{self.summarize()}')
+
+        # load adapters weights if provided
+        if hasattr(self.cfg, 'adapters_file'):
+            self.loading_base_model = False
+            with tempfile.TemporaryDirectory() as tmpdir:
+                adapters_state_dict = self.extract_state_dict_from(
+                    self.cfg.adapters_file, tmpdir, save_restore_connector=NLPSaveRestoreConnector()
+                )
+                self._load_adapters_weights(adapters_state_dict)
+                logging.info(f'Adapters weights loaded successfully from {self.cfg.adapters_file}')
+            self.loading_base_model = True
+
+    def _validate_adapters_cfg(self, cfg):
+        assert cfg.type in ['parallel_adapter', 'linear_adapter']
+        assert hasattr(cfg, 'adapter_dim')
+        assert cfg.norm_position in ['pre', 'post']
+
+        if hasattr(cfg, 'norm_type'):
+            assert cfg.norm_type in ['mixedfusedlayernorm', 'layernorm']
+
+        for val in ['row_init_method', 'column_init_method']:
+            if hasattr(cfg, val):
+                assert cfg.get(val) in ['xavier', 'zero', 'normal']
+
+    def _state_dict(self, destination=None, prefix=None, keep_vars=False):
+        state_dict_ = {}
```

Comment:
Just a note that this will break with pp>2. but this can be fixed in a later PR.